### PR TITLE
feat: Enable redesigned transaction confirmations for all users

### PR DESCRIPTION
**Pull request by pedronfigueiredo**
## **Description**

Migrates all users to the redesigned transactions. The setting will be automatically toggled on.

This change initially broke a number of end-to-end tests. After an analysis, I grouped these failing tests into three categories:

1. User flows that have to be tested before enabling the redesign to all users, 🔴

This includes multichain and transaction insights tests. All tests on these files were tweaked to pass on redesigned confirmations. Specific tests that initially broke for new confirmations were also duplicated and included the temporary helper method `tempToggleSettingRedesignedTransactionConfirmations` so they were run for the old confirmations that we still need to support.

2. User flows that we don't need to migrate immediately, but will have to be migrated or adapted when old confirmations flows are deleted, 🟡

For these tests, we simply added `tempToggleSettingRedesignedTransactionConfirmations`. Once we remove the old flows, we can modify the tests to support the new flows. We didn't think it was as urgent to test these flows with the redesigned screens.

3. User flows that we don't need to migrate to redesigned confirmations. 🟢

For these tests, we simply added `tempToggleSettingRedesignedTransactionConfirmations`. Once we remove the old confirmation screens, we can delete these tests because they are no longer relevant, or otherwise already tested in the redesigned confirmation specific tests.

### Summary of the e2e tests that broke with the migration

1. User flows that have to be tested before enabling the redesign to all users, 🔴
    - test/e2e/snaps/test-snap-txinsights-v2.spec.js ✅
    - test/e2e/snaps/test-snap-txinsights.spec.js ✅
    - test/e2e/json-rpc/switchEthereumChain.spec.js ✅
    - test/e2e/tests/request-queuing/batch-txs-per-dapp-diff-network.spec.js ✅
    - test/e2e/tests/request-queuing/batch-txs-per-dapp-extra-tx.spec.js ✅
    - test/e2e/tests/request-queuing/batch-txs-per-dapp-same-network.spec.js ✅
    - test/e2e/tests/request-queuing/dapp1-send-dapp2-signTypedData.spec.js ✅
    - test/e2e/tests/request-queuing/dapp1-switch-dapp2-eth-request-accounts.spec.js ✅
    - test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js ✅
    - test/e2e/tests/request-queuing/multi-dapp-sendTx-revokePermission.spec.js ✅
    - test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js ✅
    - test/e2e/tests/request-queuing/switch-network.spec.js ✅
    - test/e2e/tests/request-queuing/ui.spec.js ✅
    - test/e2e/tests/account/snap-account-transfers.spec.ts ✅

2. User flows that we don't need to migrate immediately, but will have to be migrated or adapted when old confirmations flows are deleted, 🟡
    - test/e2e/json-rpc/eth_sendTransaction.spec.js
    - test/e2e/tests/account/add-account.spec.ts
    - test/e2e/tests/petnames/petnames-transactions.spec.js
    - test/e2e/tests/ppom/ppom-blockaid-alert-simple-send.spec.js
    - test/e2e/tests/responsive-ui/metamask-responsive-ui.spec.js
    - test/e2e/tests/tokens/custom-token-send-transfer.spec.js
    - test/e2e/tests/transaction/change-assets.spec.js
    - test/e2e/tests/transaction/edit-gas-fee.spec.js
    - test/e2e/tests/transaction/gas-estimates.spec.js
    - test/e2e/tests/transaction/multiple-transactions.spec.js
    - test/e2e/tests/transaction/navigate-transactions.spec.js
    - test/e2e/tests/transaction/send-edit.spec.js
    - test/e2e/tests/transaction/send-eth.spec.js
    - test/e2e/tests/transaction/send-hex-address.spec.js

3. User flows that we don't need to migrate to redesigned confirmations. 🟢
    - test/e2e/tests/dapp-interactions/contract-interactions.spec.js
    - test/e2e/tests/dapp-interactions/dapp-tx-edit.spec.js
    - test/e2e/tests/dapp-interactions/failing-contract.spec.js
    - test/e2e/tests/network/network-error.spec.js
    - test/e2e/tests/settings/4byte-directory.spec.js
    - test/e2e/tests/settings/show-hex-data.spec.js
    - test/e2e/tests/tokens/custom-token-add-approve.spec.js
    - test/e2e/tests/tokens/increase-token-allowance.spec.js
    - test/e2e/tests/transaction/simple-send.spec.ts
    - test/e2e/tests/tokens/nft/erc721-interaction.spec.js
    - test/e2e/tests/tokens/nft/erc1155-interaction.spec.js
    - test/e2e/tests/tokens/nft/send-nft.spec.js

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/28321?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/3026

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by jpuri — DISMISSED** [review]

File: test/e2e/helpers.js [hunk 1]
```diff
@@ -934,6 +934,41 @@ async function tempToggleSettingRedesignedConfirmations(driver) {
...
+  // Ensure we are on the extension window
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.ExtensionInFullScreenView);
+
+  // Open settings menu button
+  await driver.clickElement('[data-testid="account-options-menu-button"]');
+
+  // fix race condition with mmi build
+  if (process.env.MMI) {
+    await driver.waitForSelector('[data-testid="global-menu-mmi-portfolio"]');
+  }
```
Comment (digiwand) [inline test/e2e/helpers.js:954]: 👍🏼

File: app/scripts/migrations/132.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+): Record<string, unknown> {
+  const preferencesControllerState = state?.PreferencesController as
+    | Record<string, unknown>
+    | undefined;
+
+  const preferences = preferencesControllerState?.preferences as
+    | Record<string, unknown>
+    | undefined;
+
+  if (isObject(preferencesControllerState) && isObject(preferences)) {
```
Comment (digiwand) [inline app/scripts/migrations/132.ts:40]: I think there's no need to check `isObject(preferencesControllerState)` since `preferences` is derived from `preferencesControllerState`
```suggestion
  if (isObject(preferences)) {
```

File: app/scripts/migrations/132.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(
+  state: Record<string, unknown>,
+): Record<string, unknown> {
+  const preferencesControllerState = state?.PreferencesController as
+    | Record<string, unknown>
+    | undefined;
```
Comment (digiwand) [inline app/scripts/migrations/132.ts:34]: We have supported the case when no preferences state is found in previous migrations. It may be a good idea to support it here as well. Then we can ensure we set `redesignedTransactionsEnabled` for this migration. Maybe it's automatically set to true by default later, in which case we wouldn't need this. 

example: migration 122

```
  if (!isObject(state.PreferencesController?.preferences)) {
    state.PreferencesController = {
      preferences: {},
    };
  }
```

**Review by digiwand — CHANGES_REQUESTED** [review]
Inquiries about the migration above before approving 🙏🏼

**Review by digiwand — DISMISSED** [review]

File: test/e2e/snaps/test-snap-txinsights.spec.js [hunk 4]
```diff
@@ -3,117 +3,214 @@ const {
...
+          await driver.waitForSelector({
+            text: 'Installed Snaps',
+            tag: 'h2',
+          });
+          // find and scroll to the transaction-insights test snap
+          const snapButton1 = await driver.findElement(
+            '#connecttransaction-insights',
+          );
+          await driver.scrollToElement(snapButton1);
+          // added delay for firefox (deflake)
```
Comment (FrederikBolding) [inline test/e2e/snaps/test-snap-txinsights.spec.js:40]: Can we preserve the formatting in these E2E tests? Looks like a bunch of whitespace was removed when duplicating?

File: test/e2e/tests/confirmations/helpers.ts [hunk 5]
```diff
@@ -35,12 +35,6 @@ export function withRedesignConfirmationFixtures(
           metaMetricsId: 'fake-metrics-id',
           participateInMetaMetrics: true,
         })
-        .withPreferencesController({
```
Comment (FrederikBolding) [inline test/e2e/tests/confirmations/helpers.ts:38]: Is `withRedesignConfirmationFixtures` still needed?

File: test/e2e/tests/account/snap-account-transfers.spec.ts [hunk 6]
```diff
@@ -33,6 +34,12 @@ describe('Snap Account Transfers @no-mmi', function (this: Suite) {
         ganacheServer?: Ganache;
       }) => {
         await loginWithBalanceValidation(driver, ganacheServer);
+
+        await tempToggleSettingRedesignedTransactionConfirmations(driver);
```
Comment (FrederikBolding) [inline test/e2e/tests/account/snap-account-transfers.spec.ts:38]: Do we not want to test these flows on the redesign as well?

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js [hunk 7]
```diff
@@ -3,166 +3,283 @@ const {
...
+            css: '.transaction-status-label',
+            text: 'Confirmed',
+          });
+        },
+      );
+    });
+  });
+
+  describe('Redesigned confirmation screens', function () {
+    it('tests tx insights v2 functionality', async function () {
```
Comment (FrederikBolding) [inline test/e2e/snaps/test-snap-txinsights-v2.spec.js:153]: Transaction insights v2 is meant to show the modal, but that hasn't been ported to the redesigned confirmations. I don't think this test is testing anything anymore

File: .circleci/scripts/test-run-e2e-timeout-minutes.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@ import { filterE2eChangedFiles } from '../../test/e2e/changedFilesUtil';
 
 const changedOrNewTests = filterE2eChangedFiles();
 
-// 20 minutes, plus 3 minutes for every changed file, up to a maximum of 30 minutes
-const extraTime = Math.min(20 + changedOrNewTests.length * 3, 30);
+// 20 minutes, plus 3 minutes for every changed file, up to a maximum of 60 minutes
+const extraTime = Math.min(20 + changedOrNewTests.length * 3, 60);
```
Comment (FrederikBolding) [inline .circleci/scripts/test-run-e2e-timeout-minutes.ts:6]: Curious why this is required?

File: test/e2e/snaps/test-snap-txinsights.spec.js (hunk 4 above)
Comment (FrederikBolding) [reply test/e2e/snaps/test-snap-txinsights.spec.js:40]: I wasn't referring to the indentation, but the lack of line breaks between any of the code in the copied section

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js (hunk 7 above)
Comment (pedronfigueiredo) [reply test/e2e/snaps/test-snap-txinsights-v2.spec.js:153]: I noticed the modal wasn't there while migrating the test, but should we keep the test as a placeholder, and assert the modal existence once it's ported?

File: test/e2e/tests/account/snap-account-transfers.spec.ts (hunk 6 above)
Comment (pedronfigueiredo) [reply test/e2e/tests/account/snap-account-transfers.spec.ts:38]: I originally categorised these tests as "User flows that we don't need to migrate immediately, but will have to be migrated or adapted when old confirmations flows are deleted, 🟡", but I'm changing that and duplicating them right away.

File: test/e2e/tests/confirmations/helpers.ts (hunk 5 above)
Comment (pedronfigueiredo) [reply test/e2e/tests/confirmations/helpers.ts:38]: Good point, I renamed it to `withTransactionEnvelopeTypeFixtures`

File: test/e2e/snaps/test-snap-txinsights.spec.js (hunk 4 above)
Comment (pedronfigueiredo) [reply test/e2e/snaps/test-snap-txinsights.spec.js:42]: Ah, got it. I'm not sure why this was changed in the first place but I not reverted it.

File: test/e2e/snaps/test-snap-txinsights-v2.spec.js [hunk 9]
```diff
@@ -3,166 +3,149 @@ const {
...
+            tag: 'button',
+          });
+          // wait for and click ok and wait for window to close
+          await driver.waitForSelector({ text: 'OK' });
+          await driver.clickElementAndWaitForWindowToClose({
+            text: 'OK',
+            tag: 'button',
+          });
+          // switch to test-snaps page
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.TestSnaps);
```
Comment (FrederikBolding) [inline test/e2e/snaps/test-snap-txinsights-v2.spec.js:69]: Lacking whitespace here too, otherwise LGTM!

File: test/e2e/snaps/test-snap-txinsights.spec.js [hunk 10]
```diff
@@ -3,117 +3,227 @@ const {
...
+          });
+
+          // wait for and click ok and wait for window to close
+          await driver.waitForSelector({ text: 'OK' });
+          await driver.clickElementAndWaitForWindowToClose({
+            text: 'OK',
+            tag: 'button',
+          });
+          // switch to test-snaps page and get accounts
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.TestSnaps);
```
Comment (FrederikBolding) [inline test/e2e/snaps/test-snap-txinsights.spec.js:76]: Unnecessary whitespace changes

**Review by FrederikBolding — DISMISSED** [review]

File: .circleci/scripts/test-run-e2e-timeout-minutes.ts (hunk 8 above)
Comment (chloeYue) [reply .circleci/scripts/test-run-e2e-timeout-minutes.ts]: Hi @pedronfigueiredo , we have the label "skip-e2e-quality-gate" that can bypass the check, it should be used with a valid reason to avoid abuse usage.
Also, we are trying to migrate to page object modal, since this PR is big enough, maybe we can address it in separate PR? thanks!

File: test/e2e/tests/dapp-interactions/contract-interactions.spec.js [hunk 11]
```diff
@@ -32,6 +33,11 @@ describe('Deploy contract and call contract methods', function () {
         );
         await unlockWallet(driver);
 
+        await tempToggleSettingRedesignedTransactionConfirmations(driver);
+
+        // Navigate to extension home screen
+        await driver.navigate(PAGES.HOME);
```
Comment (seaona) [inline test/e2e/tests/dapp-interactions/contract-interactions.spec.js:39]: Adding this step will delay significantly all tests. See vid. Could we instead, add one last step inside the `tempToggleSettingRedesignedTransactionConfirmations` to just close the settings button?

https://github.com/user-attachments/assets/95a37b5f-7ffe-4692-a65a-45a82002996d

**Review by seaona — DISMISSED** [review]

File: test/e2e/helpers.js [hunk 12]
```diff
@@ -934,6 +934,41 @@ async function tempToggleSettingRedesignedConfirmations(driver) {
...
+  const experimentalTabRawLocator = {
+    text: 'Experimental',
+    tag: 'div',
+  };
+  await driver.clickElement(experimentalTabRawLocator);
+
+  // Click redesigned transactions toggle
+  await driver.clickElement(
+    '[data-testid="toggle-redesigned-transactions-container"]',
+  );
```
Comment (seaona) [inline test/e2e/helpers.js:969]: I think that after clicking the toggle, we should close the settings, so we can proceed with any test normally, without the need of doing any extra step in the test, aside from calling this function (check my comment below)

File: test/e2e/tests/dapp-interactions/contract-interactions.spec.js (hunk 11 above)
Comment (pedronfigueiredo) [reply test/e2e/tests/dapp-interactions/contract-interactions.spec.js]: Added

**Comment by pedronfigueiredo** [comment]
Temporary PR created to check firefox and webpack jobs: https://github.com/MetaMask/metamask-extension/pull/28650

File: test/e2e/tests/account/snap-account-transfers.spec.ts [hunk 13]
```diff
@@ -15,151 +16,309 @@ import HomePage from '../../page-objects/pages/homepage';
...
-        );
-        await driver.switchToWindowWithTitle(
-          WINDOW_TITLES.ExtensionInFullScreenView,
-        );
-        const headerNavbar = new HeaderNavbar(driver);
-        await headerNavbar.check_accountLabel('SSK Account');
-
-        // send 1 ETH from snap account to account 1
-        await sendTransactionWithSnapAccount({
+  describe('Old confirmation screens', function () {
```
Comment (ccharly) [inline test/e2e/tests/account/snap-account-transfers.spec.ts:25]: Since both test the old and new `describe`'s are very similar:

```diff
1c1
<   describe('Redesigned confirmation screens', function () {
---
>   describe('Old confirmation screens', function () {
17a18,19
>           await tempToggleSettingRedesignedTransactionConfirmations(driver);
>
32c34
<           await sendRedesignedTransactionWithSnapAccount({
---
>           await sendTransactionWithSnapAccount({
35a38,39
>             gasFee: '0.000042',
>             totalFee: '1.000042',
64a69,70
>           await tempToggleSettingRedesignedTransactionConfirmations(driver);
>
79c85
<           await sendRedesignedTransactionWithSnapAccount({
---
>           await sendTransactionWithSnapAccount({
82a89,90
>             gasFee: '0.000042',
>             totalFee: '1.000042',
113a122,123
>           await tempToggleSettingRedesignedTransactionConfirmations(driver);
>
128c138
<           await sendRedesignedTransactionWithSnapAccount({
---
>           await sendTransactionWithSnapAccount({
131a142,143
>             gasFee: '0.000042',
>             totalFee: '1.000042',
```

I would prefer to have something like a `describe.each` to avoid repetition as much as we can.

So something like this maybe:
```ts
describe.each([
  {
    desc: 'Old',
    setup: async (driver: Driver) => {
      await tempToggleSettingRedesignedTransactionConfirmations(driver);
    },
    sendTransaction: async (driver: Driver) => {
      await sendRedesignedTransactionWithSnapAccount({
        driver,
        recipientAddress: DEFAULT_FIXTURE_ACCOUNT,
        amount: '1',
      });
    },
  },
  {
    desc: 'New',
    setup: async () => {
      // Nothing to setup for new confirmations.
    },
    sendTransaction: async (driver: Driver) => {
      await sendTransactionWithSnapAccount({
        driver,
        recipientAddress: DEFAULT_FIXTURE_ACCOUNT,
        amount: '1',
        gasFee: '0.000042',
        totalFee: '1.000042',
      });
    },
  },
])('$desc confirmation screens', ({ setup, sendTransaction }) => {
  it('can import a private key and transfer 1 ETH (sync flow)', async function () {
    await withFixtures(
        {
          fixtures: new FixtureBuilder().build(),
          ganacheOptions: multipleGanacheOptions,
          title: this.test?.fullTitle(),
        },
        async ({
          driver,
          ganacheServer,
        }: {
          driver: Driver;
          ganacheServer?: Ganache;
        }) => {
          await loginWithBalanceValidation(driver, ganacheServer);

          await setup(driver);

          await installSnapSimpleKeyring(driver);
          ...
          // Also use `sendTransaction` instead of `sendTransactionWithSnapAccount` and `sendRedesignedTransactionWithSnapAccount`.
          ...
    
})
```

Comment (ccharly) [reply test/e2e/tests/account/snap-account-transfers.spec.ts:27]: After discussing internally with @pedronfigueiredo, the "Old confirmation screens" are going to be removed "soon", so I'm ok with keeping the duplicate as-is, as long as we have a `TODO` here to mark this code-scope and make sure we're gonna remove it somedays.

Thanks!

```suggestion
  // TODO: Remove the old confirmations screen tests once we have fully migrated
  // to the new ones.
  describe('Old confirmation screens', function () {
```

**Review by ccharly — DISMISSED** [review]
LGTM, I only verified the files "owned" by the accounts team:
```
test/e2e/tests/account/add-account.spec.ts
test/e2e/tests/account/snap-account-transfers.spec.ts
```

**Review by jpuri — APPROVED** [review]

**Review by digiwand — APPROVED** [review]

**Review by ccharly — APPROVED** [review]

**Review by FrederikBolding — APPROVED** [review]
